Renderers : Accept both `render:` and `user:` attributes

andrewkaufman opened the pull request:
I'm anticipating to need a rebase after https://github.com/GafferHQ/gaffer/pull/4461 is merged, so I've made this a draft and haven't bothered updating Changes yet.

Via ImageEngine/cortex#1188 we will start loading USD constant primvars as `render:` attributes in Gaffer, so we need to pass them on to the renderer backends. Note I've only tested GafferArnold here as I don't have GafferDelight or GafferAppleseed available.

I did also update GafferScene (UI & Tests) to use the `render:` prefix for attributes, which isn't strictly necessary, but it seems better to keep the tests showing best practices rather than sticking with the soon-to-be-deprecated prefix.

However, I have a couple outstanding GafferArnoldTest issues:
- ddb8c6215f6033539e004add395546dbfb88e7e5 : This is a temp commit just to show the problem in a way that can be commented inline really. Should we add support for `Box2fData` attributes? Or should we bypass some specific list of globals-that-aren't-attributes?
- I'm getting several failing tests in `GafferArnoldTest.ArnoldShaderTest` todo with this change: 52fc9f35ded1e00d9b6f94a5830c509b0a8e3cdb (example below). I'm fairly certain they aren't caused by this PR though... any ideas?
```
======================================================================
FAIL: testColorParameterMetadata (GafferArnoldTest.ArnoldShaderTest.ArnoldShaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/python/GafferArnoldTest/ArnoldShaderTest.py", line 459, in testColorParameterMetadata
    self._testColorParameterMetadata()
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/python/GafferArnoldTest/ArnoldShaderTest.py", line 77, in wrapper
    self.fail( e.output )
AssertionError: Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/bin/__gaffer.py", line 168, in <module>
    result = app.run()
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/python/Gaffer/Application.py", line 110, in run
    return self.__run()
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/python/Gaffer/Application.py", line 150, in __run
    return self._run( self.parameters().getValidatedValue() )
  File "/home/andrewk/apps/gaffer/0.61.0.0dev/cent7.x86_64/cortex/10.3/gaffer/py2/apps/test/test-1.py", line 142, in _run
    testCase = unittest.defaultTestLoader.loadTestsFromName( name )
  File "/software/apps/python/2.7/cent7.x86_64/gcc/6.3.1/lib/python2.7/unittest/loader.py", line 109, in loadTestsFromName
    return self.suiteClass([parent(obj.__name__)])
  File "/software/apps/python/2.7/cent7.x86_64/gcc/6.3.1/lib/python2.7/unittest/case.py", line 191, in __init__
    (self.__class__, methodName))
ValueError: no such test method in <class 'GafferArnoldTest.ArnoldShaderTest.ArnoldShaderTest'>: wrapper


----------------------------------------------------------------------
```

PS. I'll drop 9bc4c0d65bfa7a417af2c11f709b53b97417866b once https://github.com/GafferHQ/gaffer/pull/4462 is merged.

johnhaddon commented on python/GafferSceneTest/CollectScenesTest.py:
```diff
@@ -129,23 +129,23 @@ def test( self ) :
 		options = GafferScene.CustomOptions()
-		options["options"].addChild( Gaffer.NameValuePlug( "user:test", "${collect:rootName}" ) )
-		self.assertTrue( "option:user:test" in options["out"]["globals"].getValue() )
+		options["options"].addChild( Gaffer.NameValuePlug( "render:test", "${collect:rootName}" ) )
```
Just noting that these are options rather than attributes - will discuss further elsewhere.

johnhaddon commented on python/GafferSceneUI/ShufflePrimitiveVariablesUI.py:
```diff
@@ -49,7 +49,7 @@
 	An additional context variable \"${source}\" can be used on the destination plugs to insert
 	the name of each source primitive variable. For example, to prefix all primitive variables
-	with \"user:\" set the source to \"*\" and the destination to \"user:${source}\").
+	with \"render:\" set the source to \"*\" and the destination to \"render:${source}\").
```
These are primvars, where prefixes don't really matter. Might be best to clarify that by not using colon-based prefix, or by using something other than `user:` or `render:`.

andrewkaufman replied:
I _think_ a common use case here is to promote primvars as attributes via an Instancer, which is why I wrote the original docs this way. Also, I found it easier to reason about the `${source}` behaviour via some prefixed name (same reason I did that in the tests originally). Happy to change it so something different if you prefer though. This was just a search/replace `user:` for `render:` really...

johnhaddon replied:
That may have been true at some point, but the Instancer does that prefixing automatically now. I agree that a common use of `${source}` is to prefix/suffix things, so it makes sense to use that as an example. But this attribute/primvar nomenclature mapping is confusing enough already, so we need to do everything we can to avoid muddying the waters.

I'm trying to think of a primitive-variable-specific example, but the closest I'm getting is `P->Pref and N->Nref`, and those seem easier just to do explicitly. Still, seems like `${source}ref` might get the idea across?

johnhaddon commented on python/GafferUI/ShufflePlugValueWidget.py:
```diff
@@ -125,7 +125,7 @@ def _updateFromPlug( self ) :
 	"""
 	The name(s) of the destination data to be created. Use \"${source}\" to insert the name of the source data.\n
-	For example, to prefix all data with \"user:\" set the destination to \"user:${source}\").
+	For example, to prefix all data with \"render:\" set the destination to \"render:${source}\").
```
Again, I'm not sure `user:` _or_ `render:` make sense in this context - might be better to use something completely generic.

johnhaddon reviewed (commented):
> ddb8c62 : This is a temp commit just to show the problem in a way that can be commented inline really. Should we add support for Box2fData attributes? Or should we bypass some specific list of globals-that-aren't-attributes?

There are a few ways of looking at this :

- This isn't an attribute, it's an option. So we can just go back to the original code, no worries.
- Options should follow the same convention as attributes. This raises a problem, because we've already been using `render:*` to mean "something that Gaffer will use to control rendering, but which is not passed to the renderer". In StandardAttributes we've been using a `gaffer:` prefix for a similar purpose. Ugh ugh ugh. I should have thought of this.

I'm not sure what to do here. Perhaps we sweep options under the rug until we do the great-attribute-rename-of-2022? Then we could pick a new prefix for the existing `render:` options.

Speaking of the great attribute rename, this PR and the Cortex PR place us in a bit of an uncanny valley where we'll be loading USD attributes like `primvars:arnold:subdiv_iterations` as `render:arnold:subdiv_iterations` and passing them to Arnold as `arnold:subdiv_iterations`, where they will do precisely _nothing_. Is that just something we're going to live with until we add the remapping for those?

> I'm getting several failing tests in GafferArnoldTest.ArnoldShaderTest todo with this change: 52fc9f3 (example below). I'm fairly certain they aren't caused by this PR though... any ideas?

Those tests now launch `gaffer` in a subprocess, so my guess would be some sort of environment problem...

andrewkaufman commented:
> This isn't an attribute, it's an option. So we can just go back to the original code, no worries.

I'd thought it was a global attribute for some reason.... I guess I saw I was in `ArnoldGlobals` and lost track of the fact it was the `options` method... What is the use-case of a `user:` option?

I do think eliminating all uses of `user:` would be better for "users" long term, and your idea of changing these few special-case options use `gaffer:` instead of `render:` or `user:` is a good one for the future.

> This PR and the Cortex PR place us in a bit of an uncanny valley where we'll be loading USD attributes like `primvars:arnold:subdiv_iterations` as `render:arnold:subdiv_iterations` ..., where they will do precisely nothing. Is that just something we're going to live with until we add the remapping for those?

I was expecting us to just live with that until Dan's USDShade efforts are ready to go. In practice, at IE, we don't ever store arnold attributes outside of Gaffer anyhow... And anyone "in the wild" who is doing so would already have a "wrongly named attribute that does nothing", so its not a particularly disruptive for anyone to just wait it out for Gaffer 0.62 I think...

> My assumption is that we need to strip off the `render:` prefix, to match the USD behaviour whereby `primvars:foo` becomes a primvar called `foo` as far as the renderer is concerned.

I'd thought this too would wait for Dan to address. This one _is_ a particularly disruptive change.... all Lookdev at IE (and elsewhere I'd think) would break immediately, unless you're suggesting we only strip `render:` and keep `user:`? Even with that, its a hard update to swallow in production as it would delay our transition to authoring `render:` attributes into the files. Maybe we need some kind of opt-out behaviour (env var?) for the prefix stripping?

Let me know what you think, I've haven't started addressing comments yet, didn't have time to actually type code today...

johnhaddon commented:
> I'd thought it was a global attribute for some reason....

Global attributes are pre-resolved on the Gaffer side as part of the general attribute flattening, and are passed through `Renderer::attributes()` like all the others.

> What is the use-case of a user: option?

Same as for an attribute - query it from a shader. Confusingly, OSL uses `getattribute( "global", "<name>" )` for this, as well as using `getattribute()` for querying primitive variables and attributes :)

> I do think eliminating all uses of user: would be better for "users" long term, and your idea of changing these few special-case options use gaffer: instead of render: or user: is a good one for the future.

Cool. As long as we know where we're heading, I'm happy to be making incremental steps in the right direction. I think we can leave options as they were for now.

>> This PR and the Cortex PR place us in a bit of an uncanny valley where we'll be loading USD attributes like primvars:arnold:subdiv_iterations as render:arnold:subdiv_iterations ..., where they will do precisely nothing. Is that just something we're going to live with until we add the remapping for those?

> I was expecting us to just live with that until Dan's USDShade efforts are ready to go. In practice, at IE, we don't ever store arnold attributes outside of Gaffer anyhow... And anyone "in the wild" who is doing so would already have a "wrongly named attribute that does nothing", so its not a particularly disruptive for anyone to just wait it out for Gaffer 0.62 I think...

OK, cool. I'm aiming for that to land in `0.61.x` rather than `0.62` though - do you foresee any problems there?

>> My assumption is that we need to strip off the render: prefix, to match the USD behaviour whereby primvars:foo becomes a primvar called foo as far as the renderer is concerned.

> I'd thought this too would wait for Dan to address. This one is a particularly disruptive change.... all Lookdev at IE (and elsewhere I'd think) would break immediately, unless you're suggesting we only strip render: and keep user:?

Yes, I'm suggesting we strip `render:` and keep `user:`. That's both what we want for the future world where `user:` has no inherent meaning, and what we want now for backwards compatibility.

> its a hard update to swallow in production as it would delay our transition to authoring render: attributes into the files. Maybe we need some kind of opt-out behaviour (env var?) for the prefix stripping?

I don't follow. I think we need to chat about this - stripping the `render:` prefix strikes me as something that needs to be done now. It's fundamental to implementing `render:` as intended. If it's hard to swallow for some reason, then `0.61` is exactly the right place for it, not `0.61.x.`

johnhaddon commented:
> stripping the render: prefix strikes me as something that needs to be done now. 

Just to clarify my thinking on this, here's one of my comments on the Cortex PR :

> Constant USD primvars are now loaded as attributes by default, which is a breaking change : they will stop rendering unless the renderer backends are updated

This is what I want to avoid, and why I expected this PR to strip the `render:` prefix.

andrewkaufman commented:
Just noting, I'm not going to have time to wrap this up before the (expected) 0.61 release date, so @johnhaddon or @danieldresser-ie will take over and address the `render:` stripping comments.

johnhaddon commented on doc/examples/lighting/lightLinkingCityAttackArnold.gfr:
```diff
@@ -732,7 +732,7 @@ __children["Assets"]["Gaffy"]["GaffyArnold"]["LookDev"]["SwitchColor"]["paramete
 __children["Assets"]["Gaffy"]["GaffyArnold"]["LookDev"]["SwitchColor"]["parameters"]["input5"].setValue( imath.Color3f( 0.129999995, 0.129999995, 0.129999995 ) )
 __children["Assets"]["Gaffy"]["GaffyArnold"]["LookDev"]["SwitchColor"]["__uiPosition"].setValue( imath.V2f( 74.058403, 108.172836 ) )
-__children["Assets"]["Gaffy"]["GaffyArnold"]["LookDev"]["InInt"]["parameters"]["name"].setValue( 'user:seed' )
+__children["Assets"]["Gaffy"]["GaffyArnold"]["LookDev"]["InInt"]["parameters"]["name"].setValue( 'render:seed' )
```
I've looked into it a bit more and I believe this is doing nothing at all - there is no source for the `seed`.

johnhaddon commented:
Closing in favour of https://github.com/GafferHQ/gaffer/pull/4470. I've also opened https://github.com/GafferHQ/gaffer/issues/4471 in an attempt to record our thought processes so far, and the work remaining to do. Please feel free to suggest clarifications there...
